File: app/src/main/java/me/zhanghai/android/files/filelist/PathObserver.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist;

import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;

import java.io.Closeable;
import java.io.IOException;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import java8.nio.file.Path;
import me.zhanghai.android.files.provider.common.PathObservable;
import me.zhanghai.android.files.provider.common.MoreFiles;

public class PathObserver implements Closeable {

    private static final long THROTTLE_INTERVAL_MILLIS = 1000;

    @Nullable
    private PathObservable mPathObservable;

    private boolean mClosed;

    @NonNull
    private final Object mLock = new Object();

    public PathObserver(@NonNull Path path, @MainThread @NonNull Runnable onChange) {
        AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> {
            synchronized (mLock) {
                if (mClosed) {
                    return;
                }
                try {
                    mPathObservable = MoreFiles.observePath(path,
                            THROTTLE_INTERVAL_MILLIS);
                } catch (UnsupportedOperationException e) {
                    // Ignored.
                    return;
                } catch (IOException e) {
                    // Ignored.
                    e.printStackTrace();
                    return;
                }
                Handler mainHandler = new Handler(Looper.getMainLooper());
                mPathObservable.addObserver(() -> mainHandler.post(onChange));
            }
        });
    }

    @Override
    @WorkerThread
    public void close() {
        AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> {
            synchronized (mLock) {
                if (mClosed) {
                    return;
                }
                mClosed = true;
                if (mPathObservable != null) {
                    try {
                        mPathObservable.close();
                    } catch (IOException e) {
                        // Ignored.
                        e.printStackTrace();
                    }
                }
            }
        });
    }
}

File: app/src/main/java/me/zhanghai/android/files/filelist/PathObserver.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist

import android.os.AsyncTask
import android.os.Handler
import android.os.Looper
import androidx.annotation.MainThread
import java8.nio.file.Path
import me.zhanghai.android.files.provider.common.PathObservable
import me.zhanghai.android.files.provider.common.observe
import me.zhanghai.android.files.util.closeSafe
import java.io.Closeable
import java.io.IOException

class PathObserver(path: Path, @MainThread onChange: () -> Unit) : Closeable {
    private var pathObservable: PathObservable? = null

    private var closed = false
    private val lock = Any()

    init {
        AsyncTask.THREAD_POOL_EXECUTOR.execute {
            synchronized(lock) {
                if (closed) {
                    return@execute
                }
                pathObservable = try {
                    path.observe(THROTTLE_INTERVAL_MILLIS)
                } catch (e: UnsupportedOperationException) {
                    // Ignored.
                    return@execute
                } catch (e: IOException) {
                    // Ignored.
                    e.printStackTrace()
                    return@execute
                }.apply {
                    val mainHandler = Handler(Looper.getMainLooper())
                    addObserver { mainHandler.post(onChange) }
                }
            }
        }
    }

    override fun close() {
        AsyncTask.THREAD_POOL_EXECUTOR.execute {
            synchronized(lock) {
                if (closed) {
                    return@execute
                }
                closed = true
                pathObservable?.closeSafe()
            }
        }
    }

    companion object {
        private const val THROTTLE_INTERVAL_MILLIS = 1000L
    }
}
